benchmark/src/benchmark-step.ts

```diff
@@ -0,0 +1,48 @@
+import type { BenchmarkValue } from './benchmark';
+import { benchmark } from './benchmark';
+import { print } from './print';
+import { Step } from './step';
+
+/**
+ * A benchmark step is time measured and run multiple times
+ * The printed out value is the time it took to run the benchmark
+ */
+export class BenchmarkStep<
+    State extends { [key in Name]?: BenchmarkValue<any> | undefined },
+    Name extends string & keyof State = string & keyof State,
+    Value extends NonNullable<State[Name]> extends BenchmarkValue<infer T>
+        ? T
+        : never = NonNullable<State[Name]> extends BenchmarkValue<infer T>
+        ? T
+        : never
+> extends Step<State, Name, State[Name]> {
+    constructor(
+        name: Name,
+        /**
+         * Will be run multiple times and must therefore be deterministic
+         */
+        private readonly functionToBenchmark: (state: State) => Value
+    ) {
+        super(name, true);
+    }
+
+    protected runStep(stepState: State) {
+        print(`  ${this.name}:`.padEnd(30, ' '));
+        const endResult = benchmark<Value>(
```

Comment:
```suggestion
        const endResult = benchmark(
```

I don't think you have to specify this.

Reply:
`// TS will error during transpiling with "Type parameter 'Value' has a circular constraint.", if we omit the generic type here` this does not happen to me here. Can you double-check, please? But it's also not important.

Reply:
I double-checked and you are correct.

I guess the following happened:

I tried both changes at once and got this error:

```ts
C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/benchmark-step.ts:14:19 - error TS2313: Type parameter 'Value' has a circular constraint.

14     Value extends NonNullable<State[Name]> extends BenchmarkValue<infer T>
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
15         ? T
   ~~~~~~~~~~~
16         : never = NonNullable<State[Name]> extends BenchmarkValue<infer T>
   ~~~~~~~~~~~~~~~

  src/benchmark-step.ts:33:27
     33         const endResult = benchmark(() => this.functionToBenchmark(stepState), {
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     34             onAfterIteration: (benchmarkValue) =>
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    ...
     37                 print('    Not deterministic!', 'red'),
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     38         });
        ~~~~~~~~~~
    Circularity originates in type at this location.

    at createTSError (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\index.ts:859:12)
    at reportTSError (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\index.ts:863:19)
    at getOutput (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\index.ts:1077:36)
    at Object.compile (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\index.ts:1433:41)
    at transformSource (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\esm.ts:400:37)
    at C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\esm.ts:278:53
    at async addShortCircuitFlag (C:\Users\julia\Documents\Studium\Bachelorproject\digital-fuesim-manv\benchmark\node_modules\ts-node\src\esm.ts:409:15)
    at async ESMLoader.load (node:internal/modules/esm/loader:407:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:11) {
  diagnosticCodes: [ 2313 ]
}
```

I only scrolled up to the second error and thought `benchmark<Value>` was necessary.
Then I tried again and got only the first error. Therefore I thought both changes triggered this error.